packages/webpack-inline-constant-exports-plugin/index.js

```diff
@@ -0,0 +1,117 @@
+const HarmonyImportSpecifierDependency = require( 'webpack/lib/dependencies/HarmonyImportSpecifierDependency' );
+const ConstDependency = require( 'webpack/lib/dependencies/ConstDependency' );
+
+const moduleTypes = [ 'javascript/auto', 'javascript/dynamic', 'javascript/esm' ];
+
+function addConstantExport( module, name, value ) {
+	if ( ! module.constantExports ) {
+		module.constantExports = {};
+	}
+	module.constantExports[ name ] = value;
+}
+
+function removeDependency( module, dep ) {
+	module.removeDependency( dep );
+	dep.module.reasons = dep.module.reasons.filter( r => r.dependency !== dep );
+}
+
+class InlinePlugin {
+	constructor( matchers ) {
+		if ( Array.isArray( matchers ) ) {
+			this.matchers = matchers;
+		} else {
+			this.matchers = [ matchers ];
+		}
+	}
+
+	isConstantsModule( module ) {
+		return module && this.matchers.some( matcher => matcher.test( module.resource ) );
+	}
+
+	apply( compiler ) {
+		compiler.hooks.compilation.tap( 'InlinePlugin', ( compilation, { normalModuleFactory } ) => {
+			const handleParser = parser => {
+				const handleExport = ( statement, declaration ) => {
+					const { module } = parser.state;
+
+					if ( ! this.isConstantsModule( module ) ) {
+						return;
+					}
+
+					if (
+						statement.type === 'ExportNamedDeclaration' &&
+						declaration.type === 'VariableDeclaration' &&
+						declaration.kind === 'const'
+					) {
+						for ( const declarator of declaration.declarations ) {
+							if ( declarator.init.type === 'Literal' ) {
+								addConstantExport( module, declarator.id.name, declarator.init.raw );
+							}
+						}
+					}
+
+					if ( statement.type === 'ExportDefaultDeclaration' && declaration.type === 'Literal' ) {
+						addConstantExport( module, 'default', declaration.raw );
+					}
+				};
+
+				parser.hooks.exportDeclaration.tap( 'InlinePlugin', handleExport );
+				parser.hooks.exportExpression.tap( 'InlinePlugin', handleExport );
+			};
+
+			moduleTypes.forEach( moduleType => {
+				normalModuleFactory.hooks.parser.for( moduleType ).tap( 'InlinePlugin', handleParser );
+			} );
+
+			compilation.hooks.optimizeDependencies.tap( 'InlinePlugin', modules => {
+				for ( const module of modules ) {
+					const importStatementDependencies = new Map();
+					const usesNonConstantDependencies = new Map();
+					const importSpecifierDependencies = [];
+
+					for ( const dep of module.dependencies ) {
+						if ( ! this.isConstantsModule( dep.module ) ) {
+							continue;
+						}
+
+						if ( dep instanceof HarmonyImportSideEffectDependency ) {
+							importStatementDependencies.set( dep.module, dep );
+						}
+
+						if ( dep instanceof HarmonyImportSpecifierDependency ) {
+							if ( dep.module.constantExports && dep.id in dep.module.constantExports ) {
+								importSpecifierDependencies.push( dep );
+								const inlineDep = new ConstDependency(
+									[
+										'/* <inline> */ ',
+										dep.shorthand ? `${ dep.name }: ` : '',
```

Comment:
I'm guessing what's happening here is for handling renamed imports? If so, I suggest adding a comment to make that a bit clearer.

Reply:
> I'm guessing what's happening here is for handling renamed imports?

This is for object property shorthands, a tricky special case:
```js
import { FOO, BAR } from 'plans';
console.log( FOO );
console.log( { BAR } );
```
gets transformed into:
```js
console.log( 'FOO' );
console.log( { BAR: 'BAR' } );
```
See the gotcha? 🙂

Reply:
Oof, good catch!

Reply:
I was the one who got caught rather than being the catcher 😄 When I first ran the plugin, it failed miserably with the [`action-log` reducer](https://github.com/Automattic/wp-calypso/blob/804a2b5a70a24fe92ad7c98a205dbcbbc565fc9f/client/state/ui/action-log/reducer.js#L24-L36).

Fortunately, webpack has to deal with exactly the [same problem](https://github.com/webpack/webpack/blob/671cb184e34e748e7c3e1d724ee24263c18c19e9/lib/dependencies/HarmonyImportSpecifierDependency.js#L163) when resolving regular imports, so the solution and infrastructure was already there for me.